### DISCUSSION TITLE
Mixing irreps allowed?

### DISCUSSION THREAD

**Discussion by haotianteng** [e3nn/e3nn#440] [Q&A]
Hi I am wondering if mixing irreps from different edge vector a good practise? For example:

```python
edge_vec1 = pos1[src] - pos1[dest]
edge_vec2 = pos2[src] - pos2[dest]
edge_sh1 = o3.spherical_harmonics(
    l=self.irreps_sh,
    x=edge_vec1,
    normalize=False,  # here we don't normalize otherwise it would not be a polynomial
    normalization="component",
)
edge_sh2 = o3.spherical_harmonics(
    l=self.irreps_sh,
    x=edge_vec2,
    normalize=False,  # here we don't normalize otherwise it would not be a polynomial
    normalization="component",
)
edge_sh = torch.cat([edge_sh1, edge_sh2], dim=1)
tp = FullyConnectedTensorProduct(
    irreps_in1 = o3.Irreps("1x0e+1x1o+1x2e+1x3o"),
    irreps_in2 = o3.Irreps("1x0e+1x1o+1x2e+1x3o+1x0e+1x1o+1x2e+1x3o")
    irreps_out = o3.Irreps("0o + 6x0e")
)
edge_features =tp(node_features[src], edge_sh_test)

```

Would this break the equivariance?

**Comment by mariogeiger** [comment]
This seems totally fine. Why would it break equivariance?

**Comment by haotianteng** [comment]
```python
edge_sh = torch.cat([edge_sh1, edge_sh2], dim=1)
```
I am mainly concerned about the above line, which concatenates sh from two different vectors and then treats it as it is from the same irreps:
```python
irreps_in2 = o3.Irreps("1x0e+1x1o+1x2e+1x3o+1x0e+1x1o+1x2e+1x3o")
```
From previous issues https://github.com/e3nn/e3nn/discussions/347#discussioncomment-3045034 , I think it mentioned that
>We use e3nn.o3.Linear layers for general
equivariant linear layers, which can only mix channels within the same
irrep.

So is it okay if the channels from the same irrep are a concatenation from irreps of different nodes?

**Reply by mariogeiger** [reply]
Yes this is totally fine
